pyrs/core/summary_generator_stress.py

```diff
@@ -0,0 +1,200 @@
+
+        def _write_summary_csv_body(handle):
+
+            def _write_number(number) -> str:
+                if math.isnan(number):
+                    return ', '
+                return str(number) + ', '
+
+            def _write_field_3d(row: int, field: str):
+                entries = ''
+                for direction in SummaryGeneratorStress.directions:
+                    if field == 'Strain':
+                        # TODO add check for strain?
+                        strain = self._get_strain_field(direction)
+                        assert(isinstance(strain, StrainField))
+
+                        strain_value = strain.values[row]
+                        strain_error = strain.errors[row]
+                        entries += _write_number(strain_value) + _write_number(strain_error)
+
+                    elif field == 'Stress':
+                        self._stress.select(direction)
+                        stress_value = self._stress.values[row]
+                        stress_error = self._stress.errors[row]
+                        entries += _write_number(stress_value) + _write_number(stress_error)
+
+                    else:
+                        peak_collection = self._get_peak_collection(direction)
+                        assert(isinstance(peak_collection, PeakCollection))
+
+                        if field == 'd':
+                            d_value = peak_collection.get_dspacing_center()[0][row]
+                            d_error = peak_collection.get_dspacing_center()[1][row]
+                            entries += _write_number(d_value) + _write_number(d_error)
+
+                        elif field == 'FWHM':
+                            fwhm_value = peak_collection.get_effective_params()[0]['FWHM'][row]
+                            fwhm_error = peak_collection.get_effective_params()[1]['FWHM'][row]
+                            entries += _write_number(fwhm_value) + _write_number(fwhm_error)
+
+                        elif field == 'Peak_Height':
+                            height_value = peak_collection.get_effective_params()[0]['Height'][row]
+                            height_error = peak_collection.get_effective_params()[1]['Height'][row]
+                            entries += _write_number(height_value) + _write_number(height_error)
+
+                return entries
+
+            # Function starts here
+            body = ''
+            for row, coordinate in enumerate(self._stress.coordinates):
+
+                line = str(coordinate[0]) + ', ' + str(coordinate[1]) + ', ' + str(coordinate[2]) + ', '
+
+                # d0 doesn't depend on direction so just picking the first peak_collection
+                for direction in SummaryGeneratorStress.directions:
+                    peak_collection = self._get_peak_collection(direction)
+                    if not peak_collection:
+                        continue
+                    d0_value = peak_collection.get_d_reference()[0][row]
```

Comment:
This assumes that `strain.coordinates` and `peak_collection.get_d_reference()[0]` are arrays of same size and same ordering. This is not generally true, even for strains containing only one peak collection. 

Please insert a `# TODO` comment stating the caveat that this algorithm will work only for cases when we have one peak collection per direction _and_ the sample points are identical for all three directions.

Once Pete is done, we'll be able to provide the mechanism to stablish the correspondence between `strain.coordinates` and `peak_collection.get_d_reference()[0]`

Reply:
thanks for clarifying these rules. I might as well do bounds checks.